File: chapters/chp1_code/vector_operations.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

typedef struct {
    double x, y, z;
} Vector3D;

Vector3D create_vector(double x, double y, double z) {
    Vector3D v = {x, y, z};
    return v;
}

double vector_magnitude(Vector3D v) {
    return sqrt(v.x*v.x + v.y*v.y + v.z*v.z);
}

Vector3D vector_normalize(Vector3D v) {
    double mag = vector_magnitude(v);
    if (mag == 0) return create_vector(0, 0, 0);
    return create_vector(v.x/mag, v.y/mag, v.z/mag);
}

double vector_dot(Vector3D a, Vector3D b) {
    return a.x*b.x + a.y*b.y + a.z*b.z;
}

Vector3D vector_cross(Vector3D a, Vector3D b) {
    return create_vector(
        a.y*b.z - a.z*b.y,
        a.z*b.x - a.x*b.z,
        a.x*b.y - a.y*b.x
    );
}

Vector3D vector_project(Vector3D a, Vector3D b) {
    double scalar = vector_dot(a, b) / vector_dot(b, b);
    return create_vector(scalar * b.x, scalar * b.y, scalar * b.z);
}

double vector_angle(Vector3D a, Vector3D b) {
    double cos_angle = vector_dot(a, b) / (vector_magnitude(a) * vector_magnitude(b));
    if (cos_angle > 1.0) cos_angle = 1.0;
    if (cos_angle < -1.0) cos_angle = -1.0;
    return acos(cos_angle);
}

double triple_scalar_product(Vector3D a, Vector3D b, Vector3D c) {
    return vector_dot(a, vector_cross(b, c));
}

void print_vector(Vector3D v, const char* name) {
    printf("%s: [%.3f, %.3f, %.3f]\n", name, v.x, v.y, v.z);
}

int main() {
    Vector3D v1 = create_vector(1, 2, 3);
    Vector3D v2 = create_vector(4, -1, 2);
    Vector3D v3 = create_vector(2, 3, -1);
    
    print_vector(v1, "Vector v1");
    print_vector(v2, "Vector v2");
    print_vector(v3, "Vector v3");
    
    double dot_product = vector_dot(v1, v2);
    Vector3D cross_product = vector_cross(v1, v2);
    Vector3D projection = vector_project(v1, v2);
    double angle = vector_angle(v1, v2);
    double triple = triple_scalar_product(v1, v2, v3);
    
    printf("\nOperations:\n");
    printf("Dot product v1·v2: %.3f\n", dot_product);
    print_vector(cross_product, "Cross product v1×v2");
    print_vector(projection, "Projection of v1 onto v2");
    printf("Angle between v1 and v2: %.3f degrees\n", angle * 180.0 / M_PI);
    printf("Triple scalar product: %.3f\n", triple);
    printf("Volume of parallelepiped: %.3f\n", fabs(triple));
    
    return 0;
}

File: chapters/chp1_code/vector_operations.rb
#!/usr/bin/env ruby

class Vector3D
  attr_accessor :x, :y, :z
  
  def initialize(x = 0, y = 0, z = 0)
    @x, @y, @z = x.to_f, y.to_f, z.to_f
  end
  
  def magnitude
    Math.sqrt(@x**2 + @y**2 + @z**2)
  end
  
  def normalize
    mag = magnitude
    return Vector3D.new(0, 0, 0) if mag == 0
    Vector3D.new(@x / mag, @y / mag, @z / mag)
  end
  
  def dot(other)
    @x * other.x + @y * other.y + @z * other.z
  end
  
  def cross(other)
    Vector3D.new(
      @y * other.z - @z * other.y,
      @z * other.x - @x * other.z,
      @x * other.y - @y * other.x
    )
  end
  
  def project_onto(other)
    scalar = self.dot(other) / other.dot(other)
    Vector3D.new(scalar * other.x, scalar * other.y, scalar * other.z)
  end
  
  def angle_with(other)
    cos_angle = self.dot(other) / (self.magnitude * other.magnitude)
    cos_angle = [[-1.0, cos_angle].max, 1.0].min  # Clamp to [-1, 1]
    Math.acos(cos_angle)
  end
  
  def to_s
    "[#{@x.round(3)}, #{@y.round(3)}, #{@z.round(3)}]"
  end
end

def triple_scalar_product(a, b, c)
  a.dot(b.cross(c))
end

# Example usage
puts "Vector Operations in Ruby"
puts "========================="

# Define three vectors
v1 = Vector3D.new(1, 2, 3)
v2 = Vector3D.new(4, -1, 2)
v3 = Vector3D.new(2, 3, -1)

puts "Vector v1: #{v1}"
puts "Vector v2: #{v2}"
puts "Vector v3: #{v3}"

# Calculate various operations
dot_product = v1.dot(v2)
cross_product = v1.cross(v2)
projection = v1.project_onto(v2)
angle = v1.angle_with(v2)
triple = triple_scalar_product(v1, v2, v3)

puts "\nOperations:"
puts "Dot product v1·v2: #{dot_product.round(3)}"
puts "Cross product v1×v2: #{cross_product}"
puts "Projection of v1 onto v2: #{projection}"
puts "Angle between v1 and v2: #{(angle * 180.0 / Math::PI).round(3)} degrees"
puts "Triple scalar product: #{triple.round(3)}"
puts "Volume of parallelepiped: #{triple.abs.round(3)}"

# Additional examples
puts "\nAdditional Examples:"

# Unit vectors
i_hat = Vector3D.new(1, 0, 0)
j_hat = Vector3D.new(0, 1, 0)
k_hat = Vector3D.new(0, 0, 1)

puts "Unit vectors:"
puts "i_hat × j_hat = #{i_hat.cross(j_hat)}"
puts "j_hat × k_hat = #{j_hat.cross(k_hat)}"
puts "k_hat × i_hat = #{k_hat.cross(i_hat)}"

# Orthogonality test
perpendicular = Vector3D.new(3, -2, 1)
test_vector = Vector3D.new(1, 1, 1)

puts "\nOrthogonality test:"
puts "#{perpendicular} · #{test_vector} = #{perpendicular.dot(test_vector)}"
puts "Vectors are #{'not ' unless perpendicular.dot(test_vector).abs < 1e-10}orthogonal"